core/src/main/java/org/apache/calcite/sql/SqlFunction.java

```diff
@@ -291,13 +291,20 @@ private RelDataType deriveType(
             }
           }
         }
+
+        // check if the identifier represents type
+        final SqlFunction x = (SqlFunction) call.getOperator();
+        final SqlIdentifier identifier = Util.first(x.getSqlIdentifier(),
+            new SqlIdentifier(x.getName(), SqlParserPos.ZERO));
+        RelDataType type = validator.getCatalogReader().getNamedType(identifier);
+        if (type != null) {
+          function = new SqlTypeConstructorFunction(identifier, type);
+          break validCoercionType;
+        }
+
         // if function doesn't exist within operator table and known function
         // handling is turned off then create a more permissive function
         if (function == null && validator.isLenientOperatorLookup()) {
-          final SqlFunction x = (SqlFunction) call.getOperator();
-          final SqlIdentifier identifier =
-              Util.first(x.getSqlIdentifier(),
-                  new SqlIdentifier(x.getName(), SqlParserPos.ZERO));
           function = new SqlUnresolvedFunction(identifier, null,
```

Comment:
Why remove these codes?

Reply:
Just refactored the code, moved it out of **if** block

Reply:
Got it.